### DISCUSSION TITLE
useradd: UID 0 is not unique

### DISCUSSION THREAD

**Discussion by jtobago** [Tecnativa/doodba#563] [Q&A]
Hello I'm trying out Doodba on ubuntu for I have heard great things about it.
I believe to have followed your video and instruction but am getting this error with the invoke git-aggregate command.

it starts building then fails on Step 2/2 useradd: UID 0 is not unique
Service 'odoo' failed to build : The command '/bin/sh -c groupadd -g $GID odoo -o     && useradd -l -md /home/odoo -s /bin/false -u $UID -g $GID odoo     && mkdir -p /var/lib/odoo     && chown -R odoo:odoo /var/lib/odoo /qa/artifacts     && chmod a=rwX /qa/artifacts     && sync' returned a non-zero code: 4

Any ideas or help would be appreciated.

**Comment by pedrobaeza** [comment]
It seems the variable $UID contains the value 0, and thus the error. Did you set GID and UID variables?

**Reply by ap-wtioit** [reply]
@jtobago the UID=0 could be implied if you are running the build with root (sudo). Are you using invoke or docker-compose build?

**Reply by jtobago** [reply]
Yes, I have used sudo invoke git-aggregate command. How do I set these variables?

**Reply by jtobago** [reply]
ok I seem to have cleared the uid error by not running as sudo. This required me to deal with a docker group issue. Now it stops after a little bit wiht the following error: `---> Using cache
 ---> Using cache
 ---> Using cache
 ---> Running in 8e6420008655
error: could not lock config file /home/odoo/.gitconfig: Permission denied
Traceback (most recent call last):
  File "/opt/odoo/common/build", line 30, in <module>
    subprocess.check_call(command)
  File "/usr/local/lib/python3.8/subprocess.py", line 364, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '/opt/odoo/common/build.d/100-repos-aggregate' returned non-zero exit status 255.
Service 'odoo' failed to build : The command '/bin/sh -c /opt/odoo/common/build && sync' returned a non-zero code: 1
oadmin@docker-odoo-dev:~/Projects/doodba$ 
`
Any Ideas why its trying to have access to .gitconfig in /home/odoo? I used a different user (oadmin) on copier.

**Reply by ap-wtioit** [reply]
This should be running in the container / build layer and with doodba there should be a /home/odoo directory existing.

It is created by the [`useradd -l -md /home/odoo -s /bin/false -u $UID -g $GID odoo`](https://github.com/Tecnativa/doodba/blob/master/16.0.Dockerfile#L166) line and [`that is before the direxec for the build directory`](https://github.com/Tecnativa/doodba/blob/master/16.0.Dockerfile#L230) runs [`100-repos-aggregate`](https://github.com/Tecnativa/doodba/blob/master/build.d/100-repos-aggregate) accessing the git config for the odoo user.

Could you add a 099-debug-permissions file to odoo/custom/build.d and make it executable (`chmod +x odoo/custom/build.d/099-debug-permissions`):
```bash
#!/usr/bin/env bash
# show the user and group of the user executing
id
# show contents of odoo home directory
ls -al /home/odoo/ || echo "/home/odoo does not exist or is not accessible"
# show contents of odoo home directory with numerical user and group info
ls -aln /home/odoo/ || echo "/home/odoo does not exist or is not accessible"
# show the odoo users id and groups
su --shell="$SHELL" odoo -c id
```
And run the build again.

**Reply by ap-wtioit** [reply]
> Yes, I have used sudo invoke git-aggregate command. How do I set these variables?

According to [`tasks_downstream.py`](https://github.com/Tecnativa/doodba-copier-template/blob/main/tasks_downstream.py#L27) you should be able to set DOODBA_GID and DOODBA_UID if they are not detected correctly:
```bash
DOODBA_GID=1000 and DOODBA_UID=1000 invoke git-aggregate
```
And make sure they match your users uid and gid.

**Reply by jtobago** [reply]
Thank you for your help so far I was not able to advance any further. I can see it now shows the uid and gid but fails on the permissions of .gitconfig. 

Any ideas to try to build.
Are there any specifics on the requirements needed for doogba using ubuntu for development?

 ---> Using cache
 ---> Running in b37b6015128f
uid=0(root) gid=0(root) groups=0(root)
total 20
drwxr-xr-x 2 root root 4096 Sep  6 04:21 .
drwxr-xr-x 1 root root 4096 Sep  6 04:21 ..
-rw-r--r-- 1 root root  220 Mar 27  2022 .bash_logout
-rw-r--r-- 1 root root 3526 Mar 27  2022 .bashrc
-rw-r--r-- 1 root root  807 Mar 27  2022 .profile
total 20
drwxr-xr-x 2 0 0 4096 Sep  6 04:21 .
drwxr-xr-x 1 0 0 4096 Sep  6 04:21 ..
-rw-r--r-- 1 0 0  220 Mar 27  2022 .bash_logout
-rw-r--r-- 1 0 0 3526 Mar 27  2022 .bashrc
-rw-r--r-- 1 0 0  807 Mar 27  2022 .profile
uid=1000(odoo) gid=1000(odoo) groups=1000(odoo)
error: could not lock config file /home/odoo/.gitconfig: Permission denied
Traceback (most recent call last):
  File "/opt/odoo/common/build", line 30, in <module>
    subprocess.check_call(command)
  File "/usr/local/lib/python3.8/subprocess.py", line 364, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '/opt/odoo/common/build.d/100-repos-aggregate' returned non-zero exit status 255.
Service 'odoo' failed to build : The command '/bin/sh -c /opt/odoo/common/build && sync' returned a non-zero code: 1
oadmin@docker-odoo-dev:~/Projects/doodba-15$

**Reply by ap-wtioit** [reply]
it looks like the home directory for the odoo user is created for the root user in your cache layer and invoke has no idea which variables have dependencies for other tasks.

when changing any parameters that are baked into the image you need to invoke a build first:
```bash
DOODBA_GID=1000 and DOODBA_UID=1000 invoke img-build
DOODBA_GID=1000 and DOODBA_UID=1000 invoke git-aggregate
```

if your development environment is also tainted you might need to start a new one (or fix the old one): https://github.com/Tecnativa/doodba-copier-template/blob/main/docs/daily-usage.md#development
